### DISCUSSION TITLE
issue: Checkbox array not behaving when inside Headless UI Dialog + Transition

### DISCUSSION THREAD

**Discussion by coryasilva** [react-hook-form/react-hook-form#9250] [Q&A] [answered]
### Version Number

7.37.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-interop-with-headless-ui-bug-forked-zcns70?file=/src/App.js

### Steps to reproduce

1. Navigate to [Codesandbox](https://codesandbox.io/s/react-hook-form-interop-with-headless-ui-bug-forked-zcns70?file=/src/App.js)
2. Click **"Open Modal with Transition"**
3. Under the section titled **"Bug: React Hook Form"**
4. Click **one or more checkboxes**
5. Click "**Submit**"
6. Observe that the submitted **data and UI is not expected**
7. _Contrast results with the correctly functioning forms title with "Sanity Check:"._

### Expected behaviour

The form and data should be the same across all three examples shown.

React hook forms + Headless UI `dialog` with `transition` misbehave. Any help would be greatly appreciated -- this one has confused me.  I opened a [ticket  with Headless UI](https://github.com/tailwindlabs/headlessui/issues/1933) and Robin kindly helped me narrow down the issue to, what seems, react-hook-forms. 

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by lastoftype** [comment]
@coryasilva It looks like your codesandbox code is missing some config for lines 43 of App.js.

Current:
```tsx
<input {...register("array")} type="checkbox" value="1" />
<input {...register("array")} type="checkbox" value="2" />
<input {...register("array")} type="checkbox" value="3" />
```

Expected:
```tsx
<input {...register("array[1]")} type="checkbox" value="1" />
<input {...register("array[2]")} type="checkbox" value="2" />
<input {...register("array[3]")} type="checkbox" value="3" />
```

In your code, `array` is the name of the input but it's overrode with each subsequent `register()` call. To fix this, you should add `array[1]` and `array[2]` and `array[3]` instead.

**Comment by coryasilva** [comment]
@lastoftype I saw that in the [docs but the corresponding example](https://react-hook-form.com/api/useform/register) led me to believe it was unnecessary.  Maybe not a bug but just a documentation clarification.

FWIW the two variations do not behave the same.

```jsx
This will unchecked values.
ex: ["1", "3"]
<input {...register("array")} type="checkbox" value="1" />
<input {...register("array")} type="checkbox" value="2" />
<input {...register("array")} type="checkbox" value="3" />
```

```jsx
This will set unchecked values as false resulting in an heterogeneous array.
ex: ["1", false, "3"]
<input {...register("array[1]")} type="checkbox" value="1" />
<input {...register("array[2]")} type="checkbox" value="2" />
<input {...register("array[3]")} type="checkbox" value="3" />
```
<img width="954" alt="Screen Shot 2022-10-19 at 10 10 45 AM" src="https://user-images.githubusercontent.com/2672972/196775034-7e465e3d-190b-404d-8073-2c0459715afe.png">

**Comment by lastoftype** [comment]
Ah ok I see now. And I checked that the example codesandbox from the docs works as expected with checkbox.

**Reply by coryasilva** [reply]
@lastoftype thanks for proving this out and giving this issue a first pass -- your answer did help me arrive at my current workaround.

**Comment by bluebill1049** [comment]
Thank you for the detailed issue report. I am not familiar with `@headlessui/react`, it's probably potential doing element mount and unmount or other operations with the actual dom which I am not sure about, which leads to issues as such. I would recommend using `Controlled` component for your usage by building a checkbox group component, or using standard react [portal](https://beta.reactjs.org/apis/react-dom#portals) API for the modal instead.

Here is a controlled component example with controller: https://codesandbox.io/s/usecontroller-checkboxes-99ld4

(Thank you @lastoftype for the help too ❤️ )

**Reply by coryasilva** [reply]
Thanks @bluebill1049.  Using a controlled component seems needlessly complicated for this use case and may result in unnecessary abstractions for simple UIs like this:

```html
<h3>Heading</h3>
<input type="checkbox" value="1"/>
<h4>Sub Heading</h4>
<input type="checkbox" value="2"/>
<input type="checkbox" value="3"/>
```

That said, it works, but I still need to clean the null values from the array.  I think I will choose to forego the controlled components and register using the array index `array.0`.  This seems to solve the problem.  I am using a zodResolver so I can clean the null values there -- it is hacky but only a few lines of code and probably more readable. 

[Example with array index on the register and an array filter at the resolver.](https://codesandbox.io/s/react-hook-form-interop-with-headless-ui-bug-with-zod-9klmw1?file=/src/App.js)

Perhaps you have some critique for this method or a simpler way or a better place to insert this logic?

**Reply by bluebill1049** [reply]
I think it's fine with dot notation syntax, the only concern is probably not following HTML standards for checkbox/radio group, but it's case by case again, if it's working for your app, I would go for that route.

**Comment by leapful** [comment, marked as answer]
@coryasilva @lastoftype @bluebill1049 

This is not an issue from react-hook-form, it's more related to Headless UI because of its handling parent - children React Fiber node.

Transition component of Headless UI is using React.cloneElement to render its child tree node

`<Transition> /* Children */ </Transition>`

will be handled by something like

```
React.cloneElement(Fragment, { children, props, refs })
```

And Transition component uses React Context to manage data like props, refs of those children therefore content inside Dialog will be rendered with snapshot value before current value, in this case it's the flag "isModalOpen".

While react-hook-form depends on real input element to be registered, switching value of `<Transition show={isModalOpen}>` performs multiple rendering on Transition's children including Dialog component that contains Form element and causes wrong registration on ref of each input name "array".

The workaround for this issue is to avoid apply isModalOpen on Transition component but inside ModalWithTransaction wrapper component.

For example:
```
const ModalWithTransition = ({ isModalOpen }) => {
    if (!isModalOpen) {
        return null; 
    }

   return (
           <Transition show appear>
                // Form elements
           </Transition>
   );
}
```

Working demo: https://codesandbox.io/s/react-hook-form-interop-with-headless-ui-bug-forked-vh2pmm?file=/src/App.js

**Reply by coryasilva** [reply]
@leapful Thank you so much for taking the time to research and **explain** this; I mostly understand but have some question that can be answered with a little reading.  

The workaround provided eliminates the exit transition so while the form works the animation doesn't.  Definitely a fair trade if I had to choose!  Let me read some source code on React.cloneElement and review again the Transition component.

**Comment by coryasilva** [comment]
Thank you all for the quick and thoughtful responses.
